llvm/docs/GitHub.rst

```diff
@@ -207,6 +207,76 @@ See more in-depth information about how to contribute in the following documenta
 * :doc:`Contributing`
 * :doc:`MyFirstTypoFix`
 
+Example Pull Request with git
+====================================
+
+Instead of using the GitHub CLI to create a PR, you can push your code to a
+remote branch on your fork and create the PR to upstream using the GitHub web
+interface.
+
+Here is an example of making a PR using git and the GitHub web interface:
+
+First follow the instructions to [fork the repository](https://docs.github.com/en/get-started/quickstart/fork-a-repo?tool=webui#forking-a-repository).
+
+Next follow the instructions to [clone your forked repository](https://docs.github.com/en/get-started/quickstart/fork-a-repo?tool=webui#cloning-your-forked-repository).
+
+Once you've cloned your forked repository,
+
+::
+
+  # Switch to the forked repo and create a new branch
+  cd llvm-project
+  git switch -c my_change
+
+  # Create your changes
+  $EDITOR file.cpp
+
+  # Don't forget clang-format
+  git clang-format
+
+  # and don't forget running your tests
+  ninja check-llvm
+
+  # Commit, use a good commit message
+  git commit file.cpp
+
+  # Push your changes to your fork branch, be mindful of
+  # your remotes here, if you don't remember what points to your
+  # fork, use git remote -v to see. Usually origin points to your
+  # fork and upstream to llvm/llvm-project
+  git push origin my_change
+
+Now back on the GitHub web interface, navigate to your fork and create a pull
+request from your branch to llvm::main.
+
+::
+
+  # If you get any review comments, come back to the branch and
+  # adjust them.
+  git switch my_change
+  $EDITOR file.cpp
+
+  # Commit your changes
+  git commit file.cpp -m "Code Review adjustments"
+
+  # Push your changes to your fork branch, be mindful of
+  # your remotes here, if you don't remember what points to your
+  # fork, use git remote -v to see. Usually origin points to your
+  # fork and upstream to llvm/llvm-project
+  git push origin my_change
+
+  # When your PR is accepted, you can now rebase it and make sure
+  # you have all the latest changes.
+  git rebase -i origin/main
```

Comment:
What is the point of the rebase here?

Reply:
Took this from above: https://github.com/llvm/llvm-project/pull/65393/files#diff-2ffdcd91784145f84fd7eba67dacb7ef2b20ea5469917531bf119e2207f77a3dL194

It allows you to run `ninja check-llvm` in a state that is synced with main + this commit. This increases the likelihood that the merging of this PR will not cause test failures.

Reply:
OK but the comment below about running ninja does not seem optional, so we should move the part about whether or not to run ninja above the rebase

Reply:
From the docs above: `After your PR is reviewed and accepted, you want to rebase your branch to ensure you won’t encounter merge conflicts when landing the PR.`

Maybe what we'd like is to move the rebase and ninja steps to be placed above pushing changes to the branch. WDYT?

Reply:
This does not make sense, you don't need to rebase to ensure you won't encounter merge conflicts, the GitHub already tells you that and **requests** you to fix it if needed!

The only reason to rebase is line 278: "If this PR is older and you get a lot of new commits with the rebase, you might want to re-run tests and make sure nothing broke."

But right now it's a "might" and it's at the end of the flow, whereas it should be the the first and foremost thing here

Reply:
I agree with you that we do not need to rebase to avoid merge conflicts unless prompted by GitHub. Maybe it would be nice if we added a step to rebase if GitHub says there are conflicts.

I concede that it may be optional to rebase when there are no conflicts, running ninja, and then merging PR. But I think it is good practice, since it will prevent failures caused by the interaction of the PR and commits on main that the PR has not seen.

Maybe we can reword the example here to note that it is not required, but good practice. WDYT?

Reply:
Recommended is fine with me either way: I was pointing just at inconsistencies in the flow.

Reply:
I have pushed an update that should resolve those inconsistencies.

Reply:
Thanks!